rfcs/0050-vm-syscalls-3/0050-vm-syscalls-3.md

```diff
@@ -12,174 +12,147 @@ Created: 2023-04-17
+int ckb_write(uint64_t fd, const void* buffer, size_t* length);
 ```
 
-- Length up to 256K.
-- If the written length is greater than the limit given by *Spawn*, the final written length is the minimum of the two.
-- Returns 0 for success, otherwise for failure.
-- This function is optional. Not every child script needs to call this, and nothing will happen if you call it from the prime script.
+Read and Write must appear in pairs; otherwise, the process will be blocked indefinitely. When using `ckb_read` and `ckb_write`, there may be the following scenarios:
 
-### Spawn example
+- If the Writer writes W bytes and the Reader reads R bytes where W > R, the Writer will block, and the Reader will return immediately with R bytes in `*length`. The Reader can then call `ckb_read` again to read the remaining W - R bytes.
+- If the Writer writes W bytes and the Reader reads R bytes where W <= R, both the Writer and the Reader will return immediately with W bytes in `*length`.
 
-Consider the creation of a dependency library with a straightforward function that receives parameters, concatenates them, and subsequently returns the resulting string to the caller.
+### Close
+[Close]: #close
 
-**lib_strcat.c**
+This syscall manually closes a file descriptor. After calling this, any attempt to read/write the file descriptor pointed to the other end would fail. After using close, there are four typical situations:
 
-```c
-#include <stdint.h>
-#include <string.h>
-
-#include "ckb_syscalls.h"
+- close writer, and then try to write data to writer through `ckb_write`. In this case `ckb_write` will fail and return error(6).
+- close writer, and then try to read data from reader through `ckb_read`. In this case if there is unread data in Pipe, ckb_read will execute normally; otherwise, it will return error(7).
+- close reader, and then try to write data to writer through `ckb_write`. In this case `ckb_write` will fail and return error(7).
+- close reader, and then try to read data from reader through `ckb_read`. In this case `ckb_read` will fail and return error(6).
 
-int main(int argc, char *argv[]) {
-  char content[80];
-  for (int i = 0; i < argc; i++) {
-    strcat(content, argv[i]);
-  }
-  uint64_t content_size = (uint64_t)strlen(content);
-  ckb_set_content(&content[0], &content_size);
-  if (content_size != (uint64_t)strlen(content)) {
-    return 1;
-  }
-  return 0;
-}
+```c
+int ckb_close(uint64_t fd);
 ```
 
-We can call this dependent library in the prime script. The prime script passes in two parameters "hello", "world" and checks if the return value is equal to "helloworld”:
+It's not always necessary to manually close file descriptors. When a process is terminated, all file descriptors owned by the process are automatically closed.
+
+### Wait
+[Wait]: #wait
 
-**prime.c**
+The syscall pauses until the execution of a process specified by `pid` has ended. If the process has already terminated when Wait is called, the method returns immediately.
```

Comment:
The method returns immediately, but with what return code? In addition, will it update `exit_code` part?